DoSomething/gladiator
# Caching of User data from Northstar.

#### What's this PR do?
This PR adds a caching layer within a system for User Repository data retrieval. There is now a dedicated user repository for when user data is fetched. The basic gist is that the system always checks the cache first and if it doesn't find the data it needs resorts to doing a full lookup.

There were a ton of little "gotchas" and other challenges, so feel free to ask me lots of questions! :dancer: 

#### How should this be manually tested?
Pull this shiz down, `composer dump-autoload`, then log in, try adding some users, changing their roles, and seeing if all updates as expected when viewing them again in the index listing or on their individual show pages.

#### Any background context you want to provide?
Instead of caching specific queries and entire collections of users which would repeat a lot of the data in the cache, the system caches each user individually, keyed by their Northstar ID. So when a single request to find a specific user is done, the `CacheUserRepository` first checks **Redis** for that user via their specified id, if none is found then it passes request off to the `DatabaseUserRepository`where it will find the user in the database, and use their ID to get their full profile data from Northstar. The returned data from Northstar is then passed back to the `CacheUserRepository`, where it is properly cached and then returned.

The extra magic :tophat: :sparkles: happens when a request is made to find a collection of users. Instead of a cached collection of that specific request, it firsts checks Redis for a a cache of the requested IDs, and then uses those IDs to find the individual users in cache and returns them as a collection, harnessing some of the multiple store/retrieve methods available in Laravel's `Cache` class. So that single user found in cache before, is the same record from cache returned in the collection. :boom: 

_[Interesting fact](http://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers)..._ URLs have a character limit (approx 2000). When using mongo IDs (each has 24 characters) that are passed as comma separated values of users for Northstar to search for, you can break the browser and make an invalid request when you combine 298 (which is what I was trying for all the seeded contestants, but ~83 is the actual max limit) or so ids into a single string and pop them in URL as part of the request :scream: So I ended up having to batch process requests to Northstar in 50 user increments!

#### What are the relevant tickets?
Fixes #53

#### Shoutouts
To Mr. @DFurnes who helped me think through of a few of these issues and provided some useful guidance! :trophy: 


---
@angaither @DFurnes @sbsmith86 @deadlybutter
